source/auth/auth.rst

```diff
@@ -259,16 +259,16 @@ Reauthentication
 
 On any operation that requires authentication, the server may raise the
 error ``ReauthenticationRequired`` (391), typically if the user's credential
-has expired.  Drivers MUST immediately attempt to force a reauthenication on
-the connection when this error is raised, and then re-attempt the operation.
+has expired.  Drivers MUST immediately attempt a reauthenication on
```

Comment:
Because we are now tracking connection status (via generation id, Access Token comparison, etc.) we are able to check (in a non-blocking manner) whether the connection has become invalid before each operation (which would be guaranteed to throw a 391 if we try to send it). We can then preemptively re-authenticate, rather than only re-authenticating after seeing the 391.

If we don't do this pre-emptive check, we will be unnecessarily sending a command that we know will fail, per connection per Access Token expiry.

Should we mandate this check, and do the tests allow for this behaviour?

Reply:
This is an optimization that can be made later, it shouldn't hold up this PR.

Reply:
Sounds good, created a ticket as you suggested here: https://jira.mongodb.org/browse/DRIVERS-2615